src/sentry/services/hybrid_cloud/__init__.py

```diff
@@ -46,13 +46,15 @@ def with_replacement(
         prev = self._singleton
         self.close()
 
-        if service:
-            self._singleton[silo_mode] = service
-            yield
-        else:
-            yield
-        self.close()
-        self._singleton = prev
+        try:
+            if service:
```

Comment:
This was actually a minor bug in my implementation, as I forgot that `contextdecorator` functions still need to receive exceptions from its scope and perform cleanup in a finally.  I fixed all `contextdecorators` I could find this way.  Fortunately this was only used in tests and had no impact beyond bad test error messages.